### ISSUE TITLE
Ironpython not working

### ISSUE THREAD

**Issue by Raffro** [microsoft/debugpy#523]
Info from vscode;
Version: 1.45.1 (system setup)
Commit: 5763d909d5f12fe19f215cbfdd29a91c0fa9208a
Date: 2020-05-14T08:27:35.169Z
Electron: 7.2.4
Chrome: 78.0.3904.130
Node.js: 12.8.1
V8: 7.8.279.23-electron.0
OS: Windows_NT x64 10.0.17763

I have the latest python extension installed

Running on ironpython 2.7.10

Content:
Unfortunately I don't see debugpy executing. When i run a python file in debug mode, the terminal executes the following:
`cd 'c:\Work\03_Git\test'; ${env:PYTHONIOENCODING}='UTF-8'; ${env:PYTHONUNBUFFERED}='1'; & 'C:\Tools\Python\Python37-32\python.exe' 'c:\Users\sak2lr\.vscode\extensions\ms-python.python-2020.6.91350\pythonFiles\ptvsd_launcher.py' '--default' '--client' '--host' 'localhost' '--port' '53733' 'c:\Work\03_Git\test\test.py'`

I assume there should not be ptvsd executed...

Anyway debugging in ironpython is not working. When I try to start a python file with ironpython (I just set 
`"pythonPath": "C:/TOOLS/IronPython 2.7/ipy.exe"` in the launch.json) the debugger get stuck.

I

**Comment by fabioz** [comment]
IronPython isn't really supported right now on debugpy.

@karthiknadig @int19h do you think we should add that support (add to ci/etc)?

**Comment by karthiknadig** [comment]
@Raffro Looks like you have `ms-python.python-2020.6.91350` which is from last year June. The latest should be `ms-python.python-2020.12.*`, and some time after june we switched to the newer debugger `debugpy`. Try the latest version and see if that helps.

/cc @luabud for the question about IronPython support.

**Comment by Raffro** [comment]
Hi @karthiknadig, I will try the latest release from https://marketplace.visualstudio.com/items?itemName=ms-python.python

For some reason, my vs code won't show up versions newer than the one which is installed on my PC. See here: 
![grafik](https://user-images.githubusercontent.com/72190365/104156088-b5a5eb80-53e8-11eb-821d-53c94a6df633.png)

**Comment by Raffro** [comment]
Update:
- I found out that nothing happens, when setting the Python environment executable in settings to the ipy.exe.
  But if it remains on python (2 or 3) and if I set the "python" argument in the launch.json I get further.
  Here is my launch.json:
`{                                                             `
    `"version": "0.2.0",                                       `
    `"configurations": [                                       `
`        {                                                     `
`            "name": "ironpython",                             `
`            "type": "python",                                 `
`            "request": "launch",                              `
`            "program": "C:\\Work\\03_Git\\mock_test\\test.py",`
`            "python": "C:\\TOOLS\\IronPython 2.7\\ipy.exe",   `
`            "console": "integratedTerminal"                   `
`        }                                                     `
`    ]                                                         `
`}                                                             `                                                          `
- The next traceback, which occurs is 
`File "c:\Users\myUser\.vscode\extensions\ms-python.python-2020.12.424452561\pythonFiles\lib\python\debugpy\_vendored\pydevd\pydevd.py", line 1594, in process_internal_commands
    int_cmd.do_it(self)
  File "c:\Users\myUser\.vscode\extensions\ms-python.python-2020.12.424452561\pythonFiles\lib\python\debugpy\_vendored\pydevd\_pydevd_bundle\pydevd_comm.py", line 531, in do_it
    InternalThreadCommand.do_it(self, dbg)
  File "c:\Users\myUser\.vscode\extensions\ms-python.python-2020.12.424452561\pythonFiles\lib\python\debugpy\_vendored\pydevd\_pydevd_bundle\pydevd_comm.py", line 499, in do_it
    self.method(dbg, *self.args, **self.kwargs)
  File "c:\Users\myUser\.vscode\extensions\ms-python.python-2020.12.424452561\pythonFiles\lib\python\debugpy\_vendored\pydevd\_pydevd_bundle\pydevd_comm.py", line 999, in internal_evaluate_expression_json
    _evaluate_response(py_db, request, result=err, error_message=err)
  File "c:\Users\myUser\.vscode\extensions\ms-python.python-2020.12.424452561\pythonFiles\lib\python\debugpy\_vendored\pydevd\_pydevd_bundle\pydevd_comm.py", line 902, in _evaluate_response
    py_db.writer.add_command(NetCommand(CMD_RETURN, 0, variables_response, is_json=True))
  File "c:\Users\myUser\.vscode\extensions\ms-python.python-2020.12.424452561\pythonFiles\lib\python\debugpy\_vendored\pydevd\_pydevd_bundle\pydevd_net_command.py", line 73, in __init__
    text = json.dumps(as_dict)
  File "C:\TOOLS\IronPython 2.7\Lib\json\__init__.py", line 244, in dumps
    return _default_encoder.encode(obj)
  File "C:\TOOLS\IronPython 2.7\Lib\json\encoder.py", line 209, in encode
    chunks = list(chunks)
  File "C:\TOOLS\IronPython 2.7\Lib\json\encoder.py", line 434, in _iterencode
    for chunk in _iterencode_dict(o, _current_indent_level):
  File "C:\TOOLS\IronPython 2.7\Lib\json\encoder.py", line 408, in _iterencode_dict
    for chunk in chunks:
  File "C:\TOOLS\IronPython 2.7\Lib\json\encoder.py", line 390, in _iterencode_dict
    yield _encoder(value)
  File "C:\TOOLS\IronPython 2.7\Lib\json\encoder.py", line 47, in py_encode_basestring_ascii
    s = s.decode('utf-8')
UnicodeDecodeError: 'unknown' codec can't decode byte 0xbb in position 1707: Die Bytes [BB] im Index 1707 können nicht von der angegebenen Codepage in Unicode übersetzt werden.`

  This is because of the line 252 in pydevd_vars.py:
  `expression = codecs.BOM_UTF8 + expression.encode('utf-8')`
  It seems that the `codecs.BOM_UTF8` can not be decoded by the ironpython function `s.decode('utf-8')`

- After fixing this, the debugger starts to run on run.py and finally executing the ironpython script, **but** the debugger hits no breaktpoints and I get the following vscode exception in the beginning:
  
![grafik](https://user-images.githubusercontent.com/72190365/104191214-8362b100-541d-11eb-90f0-96b246870e73.png)

**Comment by int19h** [comment]
Given that Python 2 has been EOL for a while now, and that IronPython is stuck at the same language version for the time being, it's hard to justify any significant amount of work to add support for it - especially when dropping support for Python 2 in general is going to happen eventually.

**Comment by Raffro** [comment]
I would be happy with a workaround locally, if anyone could help me out here...

FYI @int19h : It works with Python 2.7.2, so it is more an ironpython issue than a python 2. I hardly assume, that the problem persists, when ironpython 3 will be released....

**Comment by Raffro** [comment]
Update:
I found the point of the crash:
It comes in pydevd_comm.py in _evaluate_response when calling
`py_db.writer.add_command(NetCommand(CMD_RETURN, 0, variables_response, is_json=True))`
while the request.arguments.expression is
`if 'debugpy' not in sys.modules:`
    `sys.path.insert(0, {debugpy_dir!r})`
    `try:`
        `import debugpy`
    `finally:`
        `del sys.path[0]`

**Comment by int19h** [comment]
To clarify - we do support CPython 2.7 officially. The problem is that when we drop it and move to py3-only, we'd also have to drop any IronPython support, unless it supports py3 by then - which seems unlikely, seeing how there isn't even a tentative release date for that.

**Comment by Raffro** [comment]
It would be good to at least get the latest ironpython 2 version running. When the support is then dropped for that would be okay, because I think there won't be any new Ironpython 2 version.

**Comment by gramster** [comment]
To add to what @int19h said: usage of Python 2 is dropping fast, and IronPython is a small subset of Python 2 users, so IronPython usage overall is a small subset of a small subset of total usage. We would consider externally contributed pull requests that fixed things for IronPython but it is not something we would invest resources in ourselves when we have many other things to work on.

**Comment by fabioz** [comment]
@Raffro as a hint, you can try to take a look at the logging (which may show the actual error there).

* Open VS Code
* Select the command Extensions: Open Extensions Folder
* Locate the Python extension directory, typically of the form ms-python.python-2020.*.****
* In that directory ensure you do not have any debug*.log files, if you do, please delete them
* Go back into VS Code and modify your launch.json to add the setting "logToFile": true, see below:

```
{
    "name": "Python: Current File (Integrated Terminal)",
    "type": "python",
    "request": "launch",
    "program": "${file}",
    "console": "integratedTerminal",
    "logToFile": true
}
```

* Start debugging
* When done, go back into the extension directory and inspect the debug*.log files.

**Comment by Raffro** [comment]
@fabioz Thank you, I will have a look at it and hopefully find a solution so I can create a PR.

**Comment by Raffro** [comment]
I wanted to start debugging today but suddenly I can not start the python file with the configuration anymore. As long as the python path in launch.json is set to ipy.exe, I get the error message "The Python path in your debug configuration is invalid"

Where does this come from?

**Comment by Raffro** [comment]
I reinstalled ironpython and now it works again

**Comment by Raffro** [comment]
@fabioz I found out that the server is trying to attach to multipe sessions, which causes the debugger to fail.
What is the expected behaviour? Should it be connected only once or are there any other connections which should be established?

If I prevent the session to attach to another connection.server nothing happens

**Comment by fabioz** [comment]
@int19h I remember we had some docs on how the connections should work... do you know if they're up to date/where they are located?

**Comment by Raffro** [comment]
What I also saw from the logs is that the some of the arguments of the request are not really fitting with another. Maybe it should be so, maybe not:

`"debugAdapterPython": "C:\\tools\\common\\python\\python272\\python.exe"`
`"python": ["C:\\Program Files\\IronPython 2.7\\ipy.exe"]`
`"debugLauncherPython": "C:\\tools\\common\\python\\python272\\python.exe"`

**Comment by fabioz** [comment]
If an error here isn't really directly visible, my suggestion would actually be to try to start with the unit-tests (you can run then both with Python 2.7 and IronPython 2.7 and see the differences on the logging and maybe follow better what happens). If you can get the unit-tests to run with IronPython, presumably it should when it's integrated ;)

**Comment by Raffro** [comment]
@fabioz What do I need to do, to get the unit-tests running? Where do I find them (sorry for this question :))

**Comment by fabioz** [comment]
They're in this repository, so, the steps are something as:

In cmd shell:

```
git clone https://github.com/microsoft/debugpy.git
cd debugpy
SET PYTHONPATH=%CD%\src
pip install -r tests\requirements.txt
py.test tests\debugpy\test_breakpoints.py -k test_add_and_remove_breakpoint
```

Note that you may have to do some tweaks for things that aren't available in ironpython (probably psutil and pytest-xdist, but I haven't really tested it).

**Comment by Raffro** [comment]
@fabioz: What do you exactly mean by "py.test", this cannot be found

**Comment by fabioz** [comment]
That's a shortcut to `python -m pytest`.

**Comment by Raffro** [comment]
@fabioz How can I set the python interpreter which runs the tests?

**Comment by fabioz** [comment]
As a note, one possibility could be having the tests run with Python but having the spawned processes be run with IronPython (that way you don't need to port the tests infrastructure to IronPython, which is probably a bigger task).

**Comment by fabioz** [comment]
p.s.: Search for the `sys.executable` references to find possible places for that ;)

**Comment by int19h** [comment]
Ideally, you should be running the tests with tox. This also takes care of running them using different Python versions, and you can edit tox.ini to add IronPython. For development, I usually use an invocation similar to:
```sh
tox --develop -e py38 -- -k "test_add_and_remove_breakpoint"
```
(tox will install the prerequisites automatically and take care of PYTHONPATH, so you shouldn't need to do anything else)

**Comment by int19h** [comment]
Regarding this:
```json5
"debugAdapterPython": "C:\\tools\\common\\python\\python272\\python.exe"
"python": ["C:\\Program Files\\IronPython 2.7\\ipy.exe"]
"debugLauncherPython": "C:\\tools\\common\\python\\python272\\python.exe"
```
These correspond to Python interpreters used to spawn the various processes that together constitute debugpy. In particular, the adapter and the launcher do not use "python". That is intentional, but it shouldn't really affect you in any way, because there's no requirement for them all to use the same Python version, and the only part that *has* to use IronPython is the one that runs in the same process as the debugged code - i.e. the debug server.

**Comment by int19h** [comment]
Regarding multiple connections; we don't have a good doc for this, unfortunately, and the stuff that we [do have](https://github.com/microsoft/debugpy/blob/main/doc/Subprocess%20debugging.md) is rather out of date.

I'm not sure what you mean by "server is trying to attach to multiple sessions", though. There's a 1:1 mapping between debug servers (debugpy.server / pydevd) and debug sessions. However, in a multiprocess scenario, there's a single debug adapter managing all those servers, and each server has a separate connection to the adapter.

**Comment by Raffro** [comment]
Unofrtunately it is too much work to get the tests running with ironpython because of missing psutil and gevent. And even if it is running i don't know whether I point out the problem or not...

@int19h: From what I have seen in the debug sessions is that there is this Connection class with its method "attach_to_session". It is called on the hand from the launcher and later from the client. But when it is called from the client side, the self.server property of this connection object is already set and this causes the debugger to fail. I didn't figure out how to get deeper into this because I cannot determine the expected behaviour

**Comment by Raffro** [comment]
Another constraint is, that pydevd can not enable process tracing (Unable to load helper lib to set tracing to all threads (unsupported python vm)) because Ironpython is not able to load the pydll attach_x86.dll

**Comment by fabioz** [comment]
> Another constraint is, that pydevd can not enable process tracing (Unable to load helper lib to set tracing to all threads (unsupported python vm)) because Ironpython is not able to load the pydll attach_x86.dll

That isn't a problem per se (it just means that some features available to CPython -- such as attaching to process or tracing previously running threads in a remote attach won't be available and the `debug_this_thread()` would be needed in such cases , but that's expected, debugging in general should still be available).

**Comment by fabioz** [comment]
> Unofrtunately it is too much work to get the tests running with ironpython because of missing psutil and gevent. And even if it is running i don't know whether I point out the problem or not...

My suggestion was that you run just with CPython 2.7 and change the needed places to use `ipy.exe` instead of `sys.executable` in the test infrastructure so that you don't have to change the tests themselves (you may not be able to run all tests in `ipy.exe`, but that's ok, just focus on the `python -m pytest tests\debugpy\test_breakpoints.py -k test_add_and_remove_breakpoint` initially and get to the point where it hits a breakpoint).

**Comment by Raffro** [comment]
> 
> 
> > Unofrtunately it is too much work to get the tests running with ironpython because of missing psutil and gevent. And even if it is running i don't know whether I point out the problem or not...
> 
> My suggestion was that you run just with CPython 2.7 and change the needed places to use `ipy.exe` instead of `sys.executable` in the test infrastructure so that you don't have to change the tests themselves (you may not be able to run all tests in `ipy.exe`, but that's ok, just focus on the `python -m pytest tests\debugpy\test_breakpoints.py -k test_add_and_remove_breakpoint` initially and get to the point where it hits a breakpoint).

I tried, but it fails in the module where the psutil import is made

**Comment by Raffro** [comment]
I managed to get it running with ipy.exe as sys.executeable but it is just Spawning the watchdog and then it get stuck.

I see in the cmd shell output (the logs dont give me anything but 
`I+30.020: Dumping logs from "C:\\Work\\03_Git\\debugpy\\tests\\_logs\\2.7-64/tests/debugpy/test_breakpoints.py/test_add_and_remove_breakpoint[program-attach_connect(cli)]`
a stacktrace starting somewhere in the python2 lib folder 
`python272\lib\site-packages\execnet\gateway_base.py", line 285, in _perform_spawn reply.run()`

In the cmd shell output I see
`[gw4] node down: Not properly terminated`
`F`
`replacing crashed worker gw4`
`gw0 [2] / gw1 [2] / gw2 [2] / gw3 [2] / gw8 C / gw5 [2] / gw6 [2] / gw7 [2]  [gw6] node down: Not properly terminated`
`F`
`replacing crashed worker gw6`
`gw0 [2] / gw1 [2] / gw2 [2] / gw3 [2] / gw8 ok / gw5 [2] / gw9 ok / gw7 [2]`

**Comment by yarbcy** [comment]
Hello guys,

Do you have any update on this issue? I see the same behavior described by author.

**Comment by fabioz** [comment]
> Hello guys,
> 
> Do you have any update on this issue? I see the same behavior described by author.

There are no updates here (`IronPython` is still not supported by `debugpy`).

**Comment by gramster** [comment]
We are unlikely to ever support IronPython; usage is extremely low and it is not a priority for us. Closing as won't fix.

**Comment by cjohn-murphy** [comment]
"We are unlikely to ever support IronPython; usage is extremely low and it is not a priority for us. "
Chicken and egg. Perhaps usage is low because Visual Studio (et al) don't support it.

**Comment by int19h** [comment]
Visual Studio did support IronPython for quite a long time in the past, including debugging it. We never saw significant use, unfortunately.
